src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1195,6 +1202,27 @@ QVariant QgsOgrProvider::defaultValue( int fieldId ) const
   return resultVar;
 }
 
+QString QgsOgrProvider::defaultValueClause( int fieldIndex ) const
+{
+  return mDefaultValues.value( fieldIndex, QString() );
+}
+
+bool QgsOgrProvider::skipConstraintCheck( int fieldIndex, QgsFieldConstraints::Constraint constraint, const QVariant &value ) const
+{
+  Q_UNUSED( constraint );
+  // If the field is a fid, skip in case it's the default value
+  if ( fieldIndex == 0 && mFirstFieldIsFid )
+  {
+    qDebug() << mDefaultValues.value( fieldIndex );
+    return ! mDefaultValues.value( fieldIndex ).isEmpty();
+  }
+  else
+  {
+    // stricter check
+    return mDefaultValues.contains( fieldIndex ) && mDefaultValues.value( fieldIndex ) == value.toString() && !value.isNull();
```

Comment:
Should the last bit here read `&& value.isNull() == mDefaultValues.value( fieldIndex ).isNull()`?

Reply:
Not sure. The logic is pretty much the same of the PG provider, I think that if the value is null (the user did not enter anything in the form field) you never want to skip the constraint check (unless it's the fid ...).

Reply:
Hmmm... `mDefaultValues` contains strings, so the not null check on it would be useless.

The current default value interaction between form and provider is a bit weird (treat a `'string'` as value, unless it matches exactly the `'defaultValueExpression()'`). Unless the whole system is changed we need to keep it this way. (I'd like to go for invalid `QVariant()` to indicate default value, but that's a job for another day ;) )